benchmarks/graphics_pipeline/SphereMesh.cpp

```diff
@@ -0,0 +1,255 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "SphereMesh.h"
+
+#include <random>
+#include <numeric>
+
+using namespace ppx;
+
+// =====================================================================
+// OrderedGrid
+// =====================================================================
+
+OrderedGrid::OrderedGrid(uint32_t count, uint32_t randomSeed)
+{
+    mSizeX = static_cast<uint32_t>(std::cbrt(count));
+    mSizeY = mSizeX;
+    mSizeZ = static_cast<uint32_t>(std::ceil(count / static_cast<float>(mSizeX * mSizeY)));
+    mStep  = 10.0f;
+
+    mOrderedPointIndices.resize(count);
+    std::iota(mOrderedPointIndices.begin(), mOrderedPointIndices.end(), 0);
+
+    // Shuffle using the `mersenne_twister` deterministic random number generator to obtain
+    // the same sphere indices for a given `kMaxSphereInstanceCount`.
+    Shuffle(mOrderedPointIndices.begin(), mOrderedPointIndices.end(), std::mt19937(randomSeed));
+}
+
+float4x4 OrderedGrid::GetModelMatrix(uint32_t sphereIndex) const
+{
+    uint32_t id = mOrderedPointIndices[sphereIndex];
+    uint32_t x  = (id % (mSizeX * mSizeY)) / mSizeY;
+    uint32_t y  = id % mSizeY;
+    uint32_t z  = id / (mSizeX * mSizeY);
+
+    return glm::translate(float3(x * mStep, y * mStep, z * mStep));
+}
+
+// =====================================================================
+// SphereMesh
+// =====================================================================
+
+void SphereMesh::ApplyGrid(const OrderedGrid& grid)
+{
+    mLowInterleavedSingleSphere  = {};
+    mLowPlanarSingleSphere       = {};
+    mHighInterleavedSingleSphere = {};
+    mHighPlanarSingleSphere      = {};
+    mLowInterleaved              = {};
+    mLowPlanar                   = {};
+    mHighInterleaved             = {};
+    mHighPlanar                  = {};
+
+    mSphereCount = grid.GetCount();
+
+    CreateAllGeometries();
+    PopulateSingleSpheres();
+    PrepareFullGeometries();
+
+    // Iterate through the spheres to adjust data unique to each sphere
+    for (uint32_t sphereIndex = 0; sphereIndex < mSphereCount; sphereIndex++) {
+        WriteSpherePosition(grid, sphereIndex);
+        AppendSphereIndicesToInterleaved(sphereIndex);
+    }
+
+    // These planar index buffers are the same as the interleaved ones
+    *(mLowPlanar.GetIndexBuffer())  = *(mLowInterleaved.GetIndexBuffer());
```

Comment:
We shouldn't use this API like this. Since we are modifying the whole object we should either use SetIndexBuffer() or use the existing API to append indices and remove non-const GetIndexBuffer(). This behaviour might be difficult to catch if something goes wrong. 

On a separate note: we are adding a non const GetVertexBuffer here that will also allow this behaviour, which perhaps should be converted to Geometry SetVertexSize() and GetVertexBufferData() instead. But we can keep that as is and revisit in the future.

Reply:
OK, I specifically want to avoid calling `Append*()` many times here, and (correct me if I'm wrong) I don't think that it can be done safely with the chunk writing `AppendIndicesU32()` which requires `uint32_t*` when the buffer data is stored as `char` vector. 
So I'll go back to adding SetIndexBuffer() and remove non-const GetIndexBuffer() then?